src/Networking/NexusMods.Networking.Downloaders/Tasks/ADownloadTask.cs

```diff
@@ -0,0 +1,254 @@
+
+        var name = contentDispositionHeader.Trim('"');
+        var size = Size.From((ulong)response.Content.Headers.ContentLength.GetValueOrDefault(0));
+        return (name, size);
+    }
+    
+    protected async Task SetStatus(DownloadTaskStatus status)
+    {
+        using var tx = Connection.BeginTransaction();
+        tx.Add(PersistentState.Id, DownloaderState.Status, (byte)status);
+        
+        if (TransientState != null)
+        {
+            var report = TransientState.ActivityStatus?.MakeTypedReport();
+            if (report == null || !report.Current.HasValue)
+            {
+                tx.Add(PersistentState.Id, DownloaderState.Downloaded, Size.Zero);
+            }
+            else
+            {
+                tx.Add(PersistentState.Id, DownloaderState.Downloaded, report.Current.Value);
+            }
+        }
+        
+        var result = await tx.Commit();
+        PersistentState = result.Remap(PersistentState);
+    }
+    
+    [Reactive]
+    public DownloaderState.Model PersistentState { get; set; } = null!;
+    
+    public AbsolutePath DownloadLocation => _downloadLocation;
+
+
+    [Reactive] public Bandwidth Bandwidth { get; set; } = Bandwidth.From(0);
+
+    [Reactive] public Size Downloaded { get; set; } = Size.From(0);
+
+    [Reactive] public Percent Progress { get; set; } = Percent.Zero;
+
+    
+    /// <inheritdoc />
+    public async Task StartAsync()
+    {
+        await Resume();
+    }
+
+    /// <inheritdoc />
+    public async Task Cancel()
+    {
+        try { await CancellationTokenSource.CancelAsync(); }
+        catch (Exception) { /* ignored */ }
+        await SetStatus(DownloadTaskStatus.Cancelled);
+    }
+
+    /// <inheritdoc />
+    public async Task Suspend()
+    {
+        await SetStatus(DownloadTaskStatus.Paused);
+        try { await CancellationTokenSource.CancelAsync(); }
+        catch (Exception) { /* ignored */ }
+        
+        // Replace the token source.
+        CancellationTokenSource = new CancellationTokenSource();
+    }
+    
+    /// <inheritdoc />
+    public async Task Resume()
+    {
+        Logger.LogInformation("Starting download of {Name}", PersistentState.FriendlyName);
+        await SetStatus(DownloadTaskStatus.Downloading);
+        TransientState = new HttpDownloaderState
+        {
+            Activity = ActivityFactory.Create<Size>(IHttpDownloader.Group, "Downloading {FileName}", DownloadLocation),
+        };
+        _ = StartActivityUpdater();
+        
+        Logger.LogDebug("Dispatching download task for {Name}", PersistentState.FriendlyName);
+        await Download(DownloadLocation, CancellationTokenSource.Token);
+        UpdateActivity();
+        await SetStatus(DownloadTaskStatus.Analyzing);
+        Logger.LogInformation("Finished download of {Name} starting analysis", PersistentState.FriendlyName);
+        await AnalyzeFile();
+        await SetStatus(DownloadTaskStatus.Completed);
```

Comment:
I think this part is causing the exception on pause and cancel. The `await Download()` task is cancelled but then `AnalyzeFile` is called assuming the download has finished instead of checking whether it was cancelled, or paused first.